streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java

```diff
@@ -0,0 +1,373 @@
+                    tasksAndActionsLock.unlock();
+                }
+            }
+        }
+
+        private void clear() {
+            tasksAndActionsLock.lock();
+            restoredActiveTasksLock.lock();
+            try {
+                tasksAndActions.clear();
+                restoredActiveTasks.clear();
+            } finally {
+                tasksAndActionsLock.unlock();
+                restoredActiveTasksLock.unlock();
+            }
+            changelogReader.clear();
+            updatingTasks.clear();
+        }
+
+        private List<TaskAndAction> getTasksAndActions() {
+            final List<TaskAndAction> tasksAndActionsToProcess = new ArrayList<>(tasksAndActions);
+            tasksAndActions.clear();
+            return tasksAndActionsToProcess;
+        }
+
+        private void addTask(final Task task) {
+            if (isStateless(task)) {
+                addTaskToRestoredTasks((StreamTask) task);
+            } else {
+                updatingTasks.put(task.id(), task);
+            }
+        }
+
+        private boolean isStateless(final Task task) {
+            return task.changelogPartitions().isEmpty() && task.isActive();
+        }
+
+        private void endRestorationIfChangelogsCompletelyRead(final Task task,
+                                                              final Set<TopicPartition> restoredChangelogs) {
+            final Collection<TopicPartition> taskChangelogPartitions = task.changelogPartitions();
+            if (restoredChangelogs.containsAll(taskChangelogPartitions)) {
+                task.completeRestoration(offsetResetter);
+                addTaskToRestoredTasks((StreamTask) task);
+                updatingTasks.remove(task.id());
+            }
+        }
+
+        private void addTaskToRestoredTasks(final StreamTask task) {
+            restoredActiveTasksLock.lock();
+            try {
+                restoredActiveTasks.add(task);
+                restoredActiveTasksCondition.signalAll();
+            } finally {
+                restoredActiveTasksLock.unlock();
+            }
+        }
+    }
+
+    enum Action {
+        ADD
+    }
+
+    private static class TaskAndAction {
+        public final Task task;
+        public final Action action;
+
+        public TaskAndAction(final Task task, final Action action) {
+            this.task = task;
+            this.action = action;
+        }
+    }
+
+    private final Time time;
+    private final ChangelogReader changelogReader;
+    private final java.util.function.Consumer<Set<TopicPartition>> offsetResetter;
+    private final Queue<TaskAndAction> tasksAndActions = new LinkedList<>();
+    private final Lock tasksAndActionsLock = new ReentrantLock();
+    private final Condition tasksAndActionsCondition = tasksAndActionsLock.newCondition();
+    private final Queue<StreamTask> restoredActiveTasks = new LinkedList<>();
+    private final Lock restoredActiveTasksLock = new ReentrantLock();
```

Comment:
And also as a wild thought (maybe we've discussed about this.. I cannot remember clearly): what if we just make all three communication channels as blocking queues, instead of one blocking queue and two pairs of queue+locks? More specifically:

1) in the updater's iteration, it would just try to `drainTo` everything from the action queue.

2) in the main thread, it would just try to `drainTo` the failed and completed tasks, and we do not use a `timeout` in the `getRestoredActiveTasks` (maybe there's a reason that a blocking with timeout is preferred, just lmk).

3) the tricky part is `getAllTasks`; for that maybe we can have a single global read-write lock, and all other functions just try to grab `read` lock, and this function grabs the `write` lock.

Reply:
I used only `BlockingQueue`s in a previous version and they were not appropriate for what I wanted to do. I cannot remember all details. In `waitIfAllChangelogsCompletelyRead()` we need a condition variable because I do not want have a busy loop when we have restored all active tasks and there are no standby tasks. 
We can discuss about the condition variable with timeout in `getRestoredActiveTasks()` further down the road. This condition variable might help to not have a busy-ish loop in the main thread. But let's see when we integrate the state updater.

Reply:
SG, let's defer this discussion down the road.